### PR TITLE
feat: Ai review implementation

### PR DESCRIPTION
**Pull request by Greyeye**
## AI Impact Review for Kubechecks

Adds an AI-powered review agent that performs impact analysis of Kubernetes manifest changes. The agent uses tool-calling to reason across diffs, rendered manifests, ArgoCD cluster state, and Helm chart sources, and posts a separate review comment alongside the existing kubechecks comment.

### What it does

When enabled, kubechecks posts a separate **AI Impact Review** comment on each PR/MR alongside the existing checks comment. The agent:

- Reads the diff and rendered manifests to understand the change
- Reads the MR/PR title and description so reviews respect author intent (description capped at 2000 chars)
- Queries ArgoCD for the live state of an app's managed resources (deployments, services, HPAs, KEDA ScaledObjects, etc.) — works across any cluster ArgoCD manages without direct K8s access. Secrets are filtered out.
- Downloads and inspects Helm chart source files to detect misspelled or unrecognized values (Helm silently ignores unknown keys)
- Produces a structured review with change summary, issues found (critical/warning/info), and a recommendation (APPROVE/WARN/FLAG)
- Optionally posts inline code suggestions directly on PR files (GitHub and GitLab), gated by `--ai-review-post-suggestions`
- Updates the commit status based on the recommendation, capped by `--worst-ai-review-state`

### Architecture

- **Provider-agnostic**: supports both Anthropic (Claude) and OpenAI (GPT) via a `Provider` interface (`pkg/aiproviders`) with separate implementations
- **Agentic tool-use loop**: the LLM calls tools iteratively to gather context before producing its review (`pkg/aireview/agent.go`)
- **Parallel tool execution**: when the LLM requests multiple tools in a single turn, they execute concurrently via goroutines
- **Anthropic prompt caching enabled**: system prompt + tool defs are cached across turns (Anthropic does not enable this by default like OpenAI does)
- **Separate from processor pipeline**: runs in parallel with existing checks, posts its own aggregated comment (one comment per PR, with sections per app)
- **Multi-app aggregation**: when a PR affects multiple ArgoCD apps, individual reviews are consolidated into a single comment via a second LLM aggregation pass (`pkg/aireview/aggregator.go`); on aggregator failure, falls back to raw concatenation wrapped in `<details>` tags
- **Per-PR app cap**: at most `--ai-review-max-apps` apps are reviewed per MR/PR (default 10, hard-clamped at 50). Skipped apps are reported in the comment.
- **Distroless-compatible**: no external runtimes needed, everything runs in Go inside the kubechecks process

### Tools available to the agent

| Tool | Purpose |
|------|---------|
| `get_diff` | Unified diff (Secrets and CRDs filtered) |
| `get_rendered_manifests` | Full YAML manifests (Secrets and CRDs filtered, truncated at 100KB) |
| `get_app_info` | ArgoCD application metadata (name, namespace, cluster, project, source) |
| `query_app_resources` | List managed resources via ArgoCD `ManagedResources` API (Secrets filtered) |
| `get_app_resource` | Get the live state of a specific resource via ArgoCD (Secrets blocked) |
| `list_chart_files` | List files in a Helm chart |
| `read_chart_file` | Read `values.yaml`, templates, helpers from a chart |
| `post_suggestion` | Post inline code suggestions on PR files |
| `submit_recommendation` | Submit structured review recommendation |

### Configuration

| Env Var | Default | Description |
|---------|---------|-------------|
| `KUBECHECKS_ENABLE_AI_REVIEW` | `false` | Enable AI impact review |
| `KUBECHECKS_AI_REVIEW_PROVIDER` | `anthropic` | `anthropic` or `openai` |
| `KUBECHECKS_AI_REVIEW_MODEL` | `claude-sonnet-4-6` | Model ID |
| `KUBECHECKS_AI_REVIEW_POST_SUGGESTIONS` | `false` | Post inline code suggestions on PR/MR files |
| `KUBECHECKS_AI_REVIEW_MAX_APPS` | `10` | Max apps reviewed per MR/PR (hard-clamped at 50) |
| `KUBECHECKS_AI_REVIEW_MAX_TURNS` | `20` | Max tool-use iterations per review |
| `KUBECHECKS_AI_REVIEW_TIMEOUT` | `5m` | Timeout per review |
| `KUBECHECKS_AI_REVIEW_SYSTEM_PROMPT` | (built-in) | Custom system prompt (overrides default) |
| `KUBECHECKS_AI_REVIEW_EXTRA_INSTRUCTIONS` | | Extra instructions appended to default prompt (e.g. org-wide policies) |
| `KUBECHECKS_ANTHROPIC_API_KEY` | | Anthropic API key |
| `KUBECHECKS_CHART_CACHE_DIR` | `/tmp/kubechecks/charts` | Helm chart cache dir |
| `KUBECHECKS_WORST_AI_REVIEW_STATE` | `warning` | Max commit status severity from AI review |
| `KUBECHECKS_ENABLE_AI_DIFF_SUMMARY` | `false` | Enable legacy AI diff summary (now uses the same provider as AI review) |

### Other changes

- **Removed `pkg/aisummary`** — diff summary now uses the `aiproviders` interface (Anthropic or OpenAI), driven by the same `--ai-review-provider` / `--ai-review-model` config
- **AI diff summary** is now gated by an explicit `KUBECHECKS_ENABLE_AI_DIFF_SUMMARY` flag (previously implicit)
- **Replaced `github.com/sashabaranov/go-openai`** with the official `github.com/openai/openai-go/v3`
- **Added `github.com/anthropics/anthropic-sdk-go`** and **`helm.sh/helm/v3`** for chart downloading and extraction
- **Path-traversal hardening** on `extractHelmValues` and `buildChangedFilesContent` (filepath.Abs + containment check), and on the Helm chart cache file accessors
- **GitLab pagination**: added `PerPage: 100` paginated loops where listing was previously single-page
- **Diff engine logging**: gitops-engine `V(1)` debug noise (CRD scheme warnings) suppressed; `processing directory` lowered to TRACE
- **Caller log format**: `file:line` → `dir/file:line` for easier source navigation
- **Dropped Prometheus tool** — multi-cluster Prometheus support was too complex to land in this PR

### Bug fixes

- **App deduplication** — when both an app manifest and its parent ApplicationSet change in the same PR, the matcher and generator could produce duplicate entries for the same app. The appset-generated version (which reflects template changes from the PR) now takes precedence.
- **Suggestion dedup key** — changed from `Path:EndLine:Suggestion` to `Path:StartLine:EndLine:Suggestion` so multi-line suggestions with different start ranges aren't incorrectly deduped
- **Inline Helm values line numbers** — `src.Helm.Values` is now wrapped with `addLineNumbers()` so line references are consistent with file-based values
- **Deterministic comment output** — `buildRawReviewBody` sorts app names before iterating
- **Duplicated diffText generation** — removed redundant call
- **`getManifestsWithRetry` final backoff** fix
- **AI review skipped when no diff** — avoids spending tokens on no-op runs

### Testing

- Unit tests for the agent loop (no tools, tool call, tool error, unknown tool, max turns, parallel calls)
- Unit tests for manifest tools (CRD/Secret filtering, truncation, table-driven)
- Unit tests for Helm chart cache (file listing, reading, path-traversal protection)
- Unit tests for `buildChangedFilesContent` (path traversal, valid files, edge cases)
- Unit tests for the `pkg/checks/aireview` Checker
- Provider unit tests (`pkg/aiproviders`)
- Integration tested on Tilt with Anthropic Claude Sonnet against live PRs

### CODE REVIEW

File: pkg/server/hook_handler.go [hunk 1]
```diff
@@ -127,7 +130,7 @@ func ProcessCheckEvent(ctx context.Context, pr vcs.PullRequest, ctr container.Co
 	defer repoMgr.Cleanup()
 
 	// If we've gotten here, we can now begin running checks (or trying to)
-	cEvent := events.NewCheckEvent(pr, ctr, repoMgr, processors)
+	cEvent := events.NewCheckEvent(pr, ctr, repoMgr, processors, aiReviewChecker)
 	if err := cEvent.Process(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/server/hook_handler.go:134]: Not a bug — Go structural typing handles this. Both interfaces are identical, with a deliberate pattern to avoid import cycles.

File: docs/usage.md [hunk 2]
```diff
@@ -47,6 +54,9 @@ The full list of supported environment variables is described below:
 |`KUBECHECKS_ARGOCD_REPOSITORY_ENDPOINT`|Location of the argocd repository service endpoint.|`argocd-repo-server.argocd:8081`|
 |`KUBECHECKS_ARGOCD_REPOSITORY_INSECURE`|True if you need to skip validating the grpc tls certificate.|`true`|
 |`KUBECHECKS_ARGOCD_SEND_FULL_REPOSITORY`|Set to true if you want to try to send the full repository to ArgoCD when generating manifests.|`false`|
+|`KUBECHECKS_CHART_CACHE_DIR`|Directory for caching downloaded Helm charts for AI review.|`/tmp/kubechecks/charts`|
+|`KUBECHECKS_ENABLE_AI_DIFF_SUMMARY`|Enable OpenAI-powered diff summary. Requires openai-api-token.|`false`|
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply docs/usage.md]: fixed

File: charts/kubechecks/templates/clusterrole.yaml [hunk 3]
```diff
@@ -6,6 +6,33 @@ rules:
   - apiGroups: ['argoproj.io']
     resources: ['applications', 'appprojects', 'applicationsets', 'services']
     verbs: ['get', 'list', 'watch']
-  - apiGroups: [''] # The core API group, which is indicated by an empty string
+  - apiGroups: ['']
     resources: ['secrets']
     verbs: ['get', 'list', 'watch']
+  {{- if .Values.configMap.env.KUBECHECKS_ENABLE_AI_REVIEW }}
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply charts/kubechecks/templates/clusterrole.yaml]: fixed

File: pkg/helmchart/cache.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+// ReadFile reads a file from a cached chart.
+func (c *Cache) ReadFile(chart, version, path string) (string, error) {
+	chartDir := c.chartPath(chart, version)
+
+	// Prevent path traversal
+	fullPath := filepath.Join(chartDir, filepath.Clean(path))
+	if !strings.HasPrefix(fullPath, chartDir) {
+		return "", fmt.Errorf("invalid path: %s", path)
+	}
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/helmchart/cache.go:135]: added filepath.Separator suffix to HasPrefix check, and added explicit error when no directory found after extraction.

File: pkg/events/check.go [hunk 5]
```diff
@@ -432,3 +501,76 @@ func (ce *CheckEvent) createNote(ctx context.Context) (*msg.Message, error) {
...
+	return header + reviewBody, worstState, allSuggestions
+}
+
+// buildRawReviewBody concatenates per-app reviews without aggregation, wrapped in <details> tags.
+func buildRawReviewBody(appReviews map[string]string) string {
+	var sb strings.Builder
+	for appName, review := range appReviews {
+		fmt.Fprintf(&sb, "<details>\n<summary><code>%s</code></summary>\n\n%s\n\n</details>\n\n", appName, review)
+	}
+	return sb.String()
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/events/check.go:583]: buildRawReviewBody now sorts app names before iterating, so the comment output is deterministic across runs.

File: pkg/events/worker.go [hunk 6]
```diff
@@ -122,11 +132,182 @@ func (w *worker) processApp(ctx context.Context, app v1alpha1.Application) {
...
+
+		if strings.TrimSpace(diffText) == "" {
+			rootLogger.Debug().Caller().Str("app", appName).Msg("no manifest changes detected, skipping AI review")
+		} else {
+			aiReviewWg.Add(1)
+			go func() {
+				defer aiReviewWg.Done()
+				w.runAIReview(ctx, app, appName, k8sVersion, jsonManifests, yamlManifests, rootLogger)
+			}()
+		}
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/events/worker.go:158]: fixed f0d71232

File: pkg/events/worker.go [hunk 7]
```diff
@@ -122,11 +132,182 @@ func (w *worker) processApp(ctx context.Context, app v1alpha1.Application) {
...
+			Int("max_retries", maxManifestRetries).
+			Str("app", appName).
+			Msg("transient error getting manifests, retrying")
+
+		// Simple backoff: 2s, 4s
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(time.Duration(attempt*2) * time.Second):
+		}
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/events/worker.go]: fixed c6a67318

File: pkg/events/check.go [hunk 8]
```diff
@@ -432,3 +502,83 @@ func (ce *CheckEvent) createNote(ctx context.Context) (*msg.Message, error) {
...
+	for _, r := range ce.aiReviewResults {
+		appReviews[r.AppName] = r.Result.Details
+		worstState = pkg.WorstState(worstState, r.Result.State)
+		for _, s := range r.Suggestions {
+			key := fmt.Sprintf("%s:%d:%s", s.Path, s.EndLine, s.Suggestion)
+			if seen[key] {
+				continue
+			}
+			seen[key] = true
+			allSuggestions = append(allSuggestions, s)
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/events/check.go:549]: fixed [15ba686](https://github.com/zapier/kubechecks/pull/498/commits/15ba686bbc5865bc4a4a561a56ce32b156cf0a91)

File: pkg/checks/aireview/check.go [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+		return ""
+	}
+
+	var parts []string
+
+	// Inline values
+	if src.Helm.Values != "" {
+		parts = append(parts, "# Inline values (spec.source.helm.values)")
+		parts = append(parts, src.Helm.Values)
+	}
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/checks/aireview/check.go:351]: fixed [15ba686](https://github.com/zapier/kubechecks/pull/498/commits/15ba686bbc5865bc4a4a561a56ce32b156cf0a91)

File: pkg/checks/aireview/check.go [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+
+	// Value files — read from cloned repo if available, with line numbers
+	if request.Repo != nil && len(src.Helm.ValueFiles) > 0 {
+		for _, vf := range src.Helm.ValueFiles {
+			filePath := filepath.Join(request.Repo.Directory, src.Path, vf)
+			data, err := os.ReadFile(filePath)
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/checks/aireview/check.go:367]: fixed [15ba686](https://github.com/zapier/kubechecks/pull/498/commits/15ba686bbc5865bc4a4a561a56ce32b156cf0a91)

File: localdev/kubechecks/values.yaml [hunk 11]
```diff
@@ -31,6 +31,17 @@ configMap:
     KUBECHECKS_USE_ARCHIVE_MODE: "true"
     KUBECHECKS_ARCHIVE_CACHE_DIR: "/tmp/kubechecks/archives"
     KUBECHECKS_ARCHIVE_CACHE_TTL: "1h"
+    # AI Diff Summary (OpenAI)
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply localdev/kubechecks/values.yaml]: fix b5ad1e01

File: pkg/aireview/tools/manifests.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+// filterManifests removes CRDs from the manifest list.
+func filterManifests(manifests []string) []string {
+	var filtered []string
+	for _, m := range manifests {
+		if isCRD(m) {
+			continue
+		}
+		filtered = append(filtered, m)
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/aireview/tools/manifests.go]: fixed in f9b14b58

File: pkg/aireview/tools/manifests.go [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+	var kept []string
+	for _, section := range sections {
+		if section == "" {
+			continue
+		}
+		if strings.Contains(section, "CustomResourceDefinition") {
+			continue
+		}
+		kept = append(kept, "===== "+section)
+	}
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/aireview/tools/manifests.go]: fixed in f9b14b58

File: pkg/aireview/tools/argocd.go [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+				if kindMatch && groupMatch && nameMatch && nsMatch {
+					if r.LiveState == "" {
+						return fmt.Sprintf("Resource %s/%s found but has no live state (not yet deployed).", params.Kind, params.Name), nil
+					}
+
+					var liveObj map[string]any
+					if err := json.Unmarshal([]byte(r.LiveState), &liveObj); err != nil {
+						return fmt.Sprintf("Failed to parse live state: %s", err), nil
+					}
+					cleanObject(liveObj)
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/aireview/tools/argocd.go:169]: fixed in 01b180e8

File: pkg/aireview/tools/argocd.go [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+					"namespace": r.Namespace,
+					"name":      r.Name,
+				}
+
+				// Include live state if available
+				if r.LiveState != "" {
+					var liveObj map[string]any
+					if err := json.Unmarshal([]byte(r.LiveState), &liveObj); err == nil {
+						cleanObject(liveObj)
+						item["live"] = liveObj
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/aireview/tools/argocd.go:104]: fixed in 01b180e8

File: pkg/checks/aireview/check.go [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+
+	// Extract user-provided Helm values if available
+	helmValues := extractHelmValues(request)
+
+	// Build changed files content with line numbers for accurate suggestions
+	changedFilesContent := buildChangedFilesContent(request)
+
+	// Bundle diff, manifests, and Helm values inline so the LLM can start reviewing immediately
+	renderedManifestsText := strings.Join(request.YamlManifests, "\n---\n")
+	userPrompt := aireview.BuildUserPrompt(request.AppName, renderedDiff, renderedManifestsText, helmValues, changedFilesContent, toolNames)
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/checks/aireview/check.go:215]: secrets are filtered now

File: cmd/root.go [hunk 17]
```diff
@@ -147,6 +147,32 @@ func init() {
...
+	// ai
+	boolFlag(flags, "enable-ai-diff-summary", "Enable AI-powered diff summary. Requires openai-api-token or anthropic-api-key.")
+	boolFlag(flags, "enable-ai-review", "Enable AI-powered impact review of manifest changes.")
+	stringFlag(flags, "ai-review-provider", "AI review provider.",
+		newStringOpts().
+			withChoices("anthropic", "openai").
+			withDefault("anthropic"))
+	stringFlag(flags, "ai-review-model", "AI review model ID.",
+		newStringOpts().
+			withDefault("claude-sonnet-4-6"))
```
Comment (abhi-kapoor) [inline cmd/root.go:159]: we might have better luck not setting a default as when the model ages, the default will need to be updated each time.

Comment (Greyeye) [reply cmd/root.go:159]: You're right, the new interface is compatible with both OpenAI and Anthropic, having the Anthropic model as default may not be ideal. i'll remove the default value.

**Review by ahinh43 — APPROVED** [review]
LGTM!
